src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs

```diff
@@ -84,13 +70,42 @@ public void OnSolutionRemoved()
                 UnregisterPrimarySolution(solutionId, synchronousShutdown: false);
             }
 
+            private void RegisterPrimarySolution()
+            {
+                this.AssertIsForeground();
+                _currentSolutionId = _workspace.CurrentSolution.Id;
+                var solutionId = _currentSolutionId;
+
+                var storageLocation = _workspace.DeferredState?.ProjectTracker.GetWorkingFolderPath(_workspace.CurrentSolution);
+
+                try
+                {
+                    _session.TryInvokeAsync(
+                        nameof(IRemoteHostService.RegisterPrimarySolutionId),
+                        new object[] { solutionId, storageLocation }, CancellationToken.None).Wait(CancellationToken.None);
+                }
+                catch (OperationCanceledException)
```

Comment:
The caller passed CancellationToken.None, why are we getting cancellation? Shouldn't the error case just be a return of false since it's a Try* API? This seems like we're playing whack-a-mole here...

Reply:
I don't want the TryInvoke to blindly swallow all cancellation and return false. it will return false if it is one of our failures we defined that could happen such as users killed OOP process.

for this specific code path, only cancellation it can throw is if service hub failed (not us, but service hub itself). for those failure, we used to crash, but with long discussion, management decide not to crash but show info bar and log those issues and NFW with service hub log attached and move on.

how we move on is, creating a new cancellationToken and throw that on the fly.

you can see it here (http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Remote/JsonRpcClient.cs,9eaa564ffdacf618)

comment seems describing old behavior, now we don't rethrow, but throw new cancellation token.

Reply:
tagging @jasonmalinowski

Reply:
No, TryInvoke should not catch all exceptions. But it seems the expectation is every caller must do that so I fail to see the difference. :smile:

Why can't TryInvoke catch just that special cancellation token's cancellation and let the other pass through?

Reply:
because if TryInvoke catches it then, caller of TryInvoke need to handle it. but if it let cancellation pass through, one at the higher layer can handle cancellation same way it handles our own cancellation.

Reply:
The expectation of .NET APIs is they don't throw OperationCancelledException with a token unrelated to the one they passed in. TPL and other APIs expect this -- this is why we regularly got bit by aggregate exceptions appearing when we screwed this up. If you pass CancellationToken.None, that means you don't expect cancellation. Yes, the call can always _fail_, but that is different than cancellation. If you want to throw a different exception and then catch that exception, I'm OK with that. I still can't explain why TryInvoke() is then bool-returning!

I think your PR here argues against yourself: the return of TryInvoke is already being ignored in various places, which means people are already ignoring the result. If you did the catch once, then you'd fix all of these.

Reply:
sure, I know how cancellation is supposed to work, but unfortunately, that is not the case here. we use cancellation token as a way to escape control flow and avoid crash from service hub. it would be nice to have other way, but right now, that is the way we settled for service hub failure. hopefully, once service hub issues are all fixed, we can remove this hack, but not in current form where we still have bunch of issues service hub is trying to fix.

...

"I still can't explain why TryInvoke() is then bool-returning!"

one of case is OOP process is not there. and not all call to service hub is TryInvoke. that is just one of way. I am not sure I am following you. in this particular one, it doesn't care whether it is failed or not since it is fire and forget. other caller do care and handle when TryInvoke failed.

Reply:
@jasonmalinowski I think people are confused by try  { call(CancellationToken.None) } catch (Cancellation)

I agree that is weird, but, like I said, this is the way we choose to handle service hub failure (http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Remote/JsonRpcClient.cs,9eaa564ffdacf618). and ensure highest chance that will not bring VS down when that happen.

@jinujoseph I think I need you here. basically, @jasonmalinowski and @sharwell seems don't like the way we decide how to handle service hub failure to not crash VS.

this is done by after talking with service hub team, sri, matt to find highest chance not to crash VS. creating and throwing a cancellation token on the fly.

their object seems not on this code itself, but the fact it can throw its own cancellation token.

Reply:
Indeed, this is most confusing, in a place where we need a 100% crystal clear API. The fact we're making a fix here indicates our own "let's not crash VS" pattern isn't working how we expected. Looking at the code doing the throwing:

https://github.com/dotnet/roslyn/blob/093cf36c56feefbf081b575378db41a6279ebe51/src/VisualStudio/Core/Next/Remote/JsonRpcClient.cs#L79-L81

...has a comment implying that was a temporary decision to guard against a problem, not something we expected to maintain long term. The PR comments imply the decision was a temporary workaround, not something we thought was good design.

Reply:
yep. and we still need that temporary solution. until these are fixed.

2: Bug 517974: VS crashes when %localappdata%\IdentityService\AccountStore.json throws UnauthorizedAccessException
10: Bug 517975: VS crashes when Identity service throws BadImageFormatException
38: Bug 518035: VsixServiceDiscovery randomly fails: ExternalSettingsManager::GetScopePaths failed to initialize PkgDefManager
Bug 518533: HubClient from devenv.exe calls into VS Installer's ServiceHub controller instead of VS's own

and more..

Reply:
> The fact we're making a fix here indicates our own "let's not crash VS" pattern isn't working how we expected

that is not true. it works most of time. since all the crashes watson we used to get got all disappeared. I have no idea where you got that conclusion.

this is only one report we got since we added the cancellation approach.

Reply:
I don't understand this bit: "for this specific code path, only cancellation it can throw is if service hub failed (not us, but service hub itself)."

If that's how servicehub works, then we should be catching and handling that exception at the deepest level possible (i.e. when we actually call into servicehub, not when we're calling into our own abstraction around it).  Classes like KeepAliveSession should not leak otu this strange behavior.  But should instead wrap it and make it operate like a normal roslyn class with proper cancellation/exception semantics.